### ISSUE TITLE
Feature Request: Gradle or maven build system as in ofbiz v16

### ISSUE THREAD

**Issue by exexzian** [ilscipio/scipio-erp#13]
is gradle or maven build system in plan?

**Comment by pplx** [comment]
Hello,

We had numerous discussions on the topic. It is certainly on the table (gradle more likely than maven), but I can't say much more at this time.

One issue is we honestly did not like what ofbiz did with gradle (at least at the time of 16/17 - we may review again at 18+ if anything has changed), so we may not want to copy as-is. A second issue is that the current system has been fairly reliable (save for some old jar rebuild detection issues) and has helped us manage complex 3rd-party library dependency problems. Several of the issues we fixed very thoroughly still exist in ofbiz.

Long story short, it's a possibility, but the implementation could be like temporarily moving backward, if you can imagine.

**Comment by madppiper** [comment]
I just want to add that "Ivy" is already used by Scipio ERP, which happens to be a package manager for apache ant, our build tool (https://en.wikipedia.org/wiki/Apache_Ivy). Ivy is compatible to the maven central so you can still use the maven packages in your own projects.

**Comment by HopscotchSen** [comment]
I hope using maven build system. @pplx

**Comment by HopscotchSen** [comment]
Does the project have plans to build with spring boot?
There are a large number of entity definition XML files in the project. Do you consider replacing them with entity definition files?
@pplx @madppiper

**Comment by madppiper** [comment]
There are currently no plans to switch to spring boot. 

We are currently having some early-stage tests with our own set of Annotations that works similar to what Spring boot uses. Our hope is that they would replace the xml files in a future release. That being said, as it is early stage, there is nothing I could show you for the time being.

**Comment by HopscotchSen** [comment]
Please consider replacing XML to make it easier to read and maintain.
I can't find more information about entity definitions, such as what does the following field configuration mean?

   <field name="primaryKeyField" type="id-ne"></field>

I don't understand what "id-ne" means? Please tell me where can I find more about entity definition
Entity definitions link: https://www.scipioerp.com/community/developer/entities/
@madppiper

**Comment by madppiper** [comment]
Yes. That's the idea behind our current work. I agree with you that XML is outdated for this.

"id-ne" is a sequenced primary key.

**Comment by madppiper** [comment]
I can also recommend our IntelliJ IDEA integration: https://plugins.jetbrains.com/plugin/12108-scipio-erp-integration

It makes it a lot easier to navigate between the xml files.

**Comment by HopscotchSen** [comment]
This is a good project, but the situation is difficult to read and maintain right now. I couldn't find more information about the entity definition details, such as "id-ne" you said.
@madppiper

**Comment by madppiper** [comment]
Yeah, i understand. We are slowly adding documentation like that to the editor and we keep working on the documentation on the website as well. If you want to chime in, your contribution would be well appreciated.

**Comment by HopscotchSen** [comment]
I want to join your team and contribute to this project, but after reading the code for a few days, it is difficult to understand the relationship between the entities and the meaning of the entity fields because the entity definitions are all in the XML file and there are no comments.

**Comment by HopscotchSen** [comment]
Hi, it's been a long time. I have been following this project, but I still can't fully understand the current design and processing. Is there any new progress in replacing XML?
@madppiper

**Comment by madppiper** [comment]
Yes - still on the roadmap. Although replacing the build process with gradle is of lesser priority at the moment, as we are currently working on other bits of the architecture that will reduce large swaps of xml elsewhere :)

But it isn't like gradle isn't happening. It is. Just bigger fish to fry atm.

**Comment by madppiper** [comment]
Do you still have issues understanding the xml? If so, can i point you to our documentation? 

https://www.scipioerp.com/community/developer/entities/
https://www.scipioerp.com/community/developer/services/
https://www.scipioerp.com/community/developer/views-requests/
https://www.scipioerp.com/community/developer/architecture/components/
etc?

**Comment by HopscotchSen** [comment]
Can you provide ER diagrams or MySQL table structure files for analysis?
@madppiper

**Comment by madppiper** [comment]
Sure, check out the "literature" list at the end of this page: 

https://www.scipioerp.com/community/developer/entities/ 

Also: You can basically checkout the datamodel in its entirety via the entity viewer: https://ce.scipioerp.com/admin/control/entitymaint

**Comment by HopscotchSen** [comment]
I have read the attachment a long time ago, but I cannot visually analyze the entity relationship. Are there other relevant documents to help understand?

**Comment by madppiper** [comment]
The PDFs at the end of the document i mentioned contains visual maps for specific usecases and displays the entities that play into that usecase including their relationship. So if you look for "Product Data", you will find an entire graph just for that. 

Meanwhile, the entity viewer shows you the relationship once you select a specific entity (just click on "Reln"). If you want a dynamic graph, you can also try your luck with external tools. But remember: there are 800+ entities, all interconnected... so i think the PDFs i mentioned and our viewer are much better suited for this.

**Comment by madppiper** [comment]
This has been a long time. But we are finally migrating towards gradle for our 4.0.0 release. We will remain compatible to the old ant build structure (making it easier to port our own components), but the core product is migrating alongside a complete replacement of the xml based definition files (eecas, secas, entities, services, controller etc etc). So stay tuned - we will roll out shortly.
